Build system incompatible with Ubuntu 16.10 and lower / Compile error due to CEF API changes in newer releases.

I just wasted most of my day off trying to build this for Ubuntu 16.04.  This version of the plugin can't even be built on Ubuntu 16.04 or 18.04 (The LTS releases.)  Only the bleeding edge 18.10 can build it and I don't have that OS installed anywhere yet.

To build this plugin you MUST have:

cmake 3.8.something
gcc-8.x  (total show stopper on Ubuntu < 18.10.)

I managed to work around the cmake requirement by building cmake from source, but I draw that line at custom building all of gcc and all of the system libraries.

I'm pretty irritated right now.  Rm -rf the whole project off my system.
Please be more specific with the version dependencies that exclude 90% of the ubuntu users out here.

Thank you for your response!  I've come back to make another attempt.  It still fails, but it gets further than it did before.

I downloaded the latest 64-bit binary for cef standard distribution, followed your directions and built the wrapper library without any visible errors.

I did a git-clone of the current obs-linuxbrowser master branch.  Then I did a wget of the zip file of your patch from the email, and applied that, and verified it changed the CMakeLists.txt file.  Then I followed the cmake/make directions from the README.md on Github, mostly.  (The README.md says to issue "mkdir build; cd build" before running the cmake, but the cmake fails within the build directory because it can't find the CMakeLists.txt file.  So I "cd .." and run from there, which seems to work.  I think the README.md also needs some correction.)

The output from that build step in the obs-linuxbrowser src root is attached.
[build.log](https://github.com/bazukas/obs-linuxbrowser/files/2730865/build.log)

The gist of it is:

```
In file included from /home/joi/src/obs-linuxbrowser/src/browser/browser-app.hpp:25:0,
                 from /home/joi/src/obs-linuxbrowser/src/browser/browser-app.cpp:29:
/home/joi/src/obs-linuxbrowser/src/browser/browser-client.hpp:40:15: error: conflicting return type specified for ‘virtual bool BrowserClient::
GetViewRect(CefRefPtr<CefBrowser>, CefRect&)’
  virtual bool GetViewRect(CefRefPtr<CefBrowser> browser, CefRect& rect) OVERRIDE;
               ^
In file included from /home/joi/src/cef_binary_3.3578.1860.g36610bd_linux64/include/cef_client.h:54:0,
                 from /home/joi/src/obs-linuxbrowser/src/browser/browser-client.hpp:20,
                 from /home/joi/src/obs-linuxbrowser/src/browser/browser-app.hpp:25,
                 from /home/joi/src/obs-linuxbrowser/src/browser/browser-app.cpp:29:
/home/joi/src/cef_binary_3.3578.1860.g36610bd_linux64/include/../include/cef_render_handler.h:85:16: error:   overriding ‘virtual void CefRende
rHandler::GetViewRect(CefRefPtr<CefBrowser>, CefRect&)’
   virtual void GetViewRect(CefRefPtr<CefBrowser> browser, CefRect& rect) = 0;
                ^
```

I'll leave this stuff on my drive for further testing.  I'm willing to provide compiled deb binaries for ubuntu if you wish, if we can get this working.

Thank you again for responding.

Regarding the build dir thing, I re-read the README.md and had misinterpreted the ..  as a "insert more stuff here" ellipsis.  So I took it literally and tried using the build directory again.  This time, I got:

```
joi@tardis:~/src/obs-linuxbrowser/build$ cmake -DCEF_ROOT_DIR=~/src/cef_binary_3.3440.1803.gde363ee_linux64 ..
-- Looking for OBS Studio in prefix /usr
-- Looking for Chromium Embedded Framework in /home/joi/src/cef_binary_3.3440.1803.gde363ee_linux64
-- Configuring done
-- Generating done
-- Build files have been written to: /home/joi/src/obs-linuxbrowser
joi@tardis:~/src/obs-linuxbrowser/build$ make
make: *** No targets specified and no makefile found.  Stop.
```

So, it looks like cmake is still writing the makefile to the source root, not my cwd in that build directory.  Cmake did create a directory with structure there.   Is there still some assumption I'm missing?  For what its worth, the 3.13.2 version of cmake is first in my PATH, a leftover from my first attempts.  Perhaps the behavior of the newest cmake differs?

I've downloaded cef_binary_3.3440.1803.gde363ee_linux64, followed the directions for building that version's wrapper library, and after some fudging realized I needed to remove the CMakeCache.txt file before restarting the build of the obs-linuxbrowser part.  I had attempted a 'make clean' that seemed to work, but it didn't remove the cache.  A quick google shows there is some debate regarding how much cmake should clean up after itself.  I'm an old-school dev and haven't used cmake before, I grew up using the autotools stuff.  make clean, make realclean, and make distclean are all targets I tried.

And we have success! It has built successfully!  And it also seems to be functional.  At least, the default settings of a newly created 'Linux Browser' showing the OBS Studio website are visible within my OBS scenes.  Interestingly, it's working with the current version available from OBS directly for Ubuntu 16.04,  22.0.2-0obsproject1~xenial.  I had expected I'd need to pin the obs package to the earlier version 21.1.2, and I may find I need to do that once I start trying to actually use an Overlay.  

Is there anything else you'd like me to try?  My next step is to try using Overlays.

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,4 +1,4 @@
-cmake_minimum_required (VERSION 3.8)
+cmake_minimum_required (VERSION 3.0)
 project (obs-linuxbrowser LANGUAGES C CXX VERSION 0.5.2)
 
 set(CMAKE_BUILD_TYPE Release CACHE STRING "CMake build type")
@@ -48,16 +48,8 @@ file(COPY ${CEF_RESOURCE_DIR}/locales DESTINATION ${PLUGIN_CEF_DATA_DIRECTORY})
 
 # compilation
 set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -std=gnu99")
-
-set(LINUXBROWSER_CXX_FEATURES
-    cxx_std_11
-    cxx_auto_type
-    cxx_constexpr
-    cxx_deleted_functions
-    cxx_nullptr
-    cxx_override
-    cxx_range_for
-)
+set(CMAKE_CXX_STANDARD 11)
+set(CMAKE_CXX_STANDARD_REQUIRED true)
 
 # Ensure a consistent command chain on all systems
 set(CMAKE_CXX_EXTENSIONS false)

